File: packages/python_backend/app/utils/storage/provider_key_storage.py

import json
import os
import time
from pathlib import Path
from typing import Type, TypeVar, Dict, Union, Any
from pydantic import BaseModel, ValidationError
from app.schemas.provider_key_schemas import ProviderKey

DATA_DIR = Path(os.getcwd()) / "data" / "provider_key_storage"
ProviderKeysStorage = Dict[int, ProviderKey]
T = TypeVar("T", bound=BaseModel)

def get_provider_keys_index_path() -> Path: return DATA_DIR / "provider_keys.json"

def ensure_dir_exists(dir_path: Path) -> None:
    try: dir_path.mkdir(parents=True, exist_ok=True)
    except Exception as e: raise IOError(f"Failed to ensure directory exists: {dir_path}")

async def read_validated_json_provider_keys(
    file_path: Path,
    model: Type[ProviderKey],
    default_value: Dict
) -> ProviderKeysStorage:
    try:
        ensure_dir_exists(file_path.parent)
        if not file_path.exists(): return default_value.copy()

        with open(file_path, "r", encoding="utf-8") as f: json_data = json.load(f)
        
        if not isinstance(json_data, dict): return default_value.copy()

        validated_data: ProviderKeysStorage = {}
        for k_str, v_data in json_data.items():
            try:
                k_int = int(k_str)
                validated_data[k_int] = model.model_validate(v_data)
            except (ValidationError, ValueError): pass
        return validated_data

    except FileNotFoundError: return default_value.copy()
    except json.JSONDecodeError: return default_value.copy()
    except Exception as e: raise IOError(f"Failed to read/parse JSON: {file_path}")

async def write_validated_json_provider_keys(
    file_path: Path,
    data: ProviderKeysStorage,
) -> ProviderKeysStorage:
    try:
        ensure_dir_exists(file_path.parent)
        json_to_write = {str(k): v.model_dump(mode="json") for k, v in data.items()}
        json_string = json.dumps(json_to_write, indent=2)
        with open(file_path, "w", encoding="utf-8") as f: f.write(json_string)
        return data
    except Exception as e: raise IOError(f"Failed to write JSON: {file_path}")

class ProviderKeyStorageUtil:
    async def read_provider_keys(self) -> ProviderKeysStorage:
        return await read_validated_json_provider_keys(get_provider_keys_index_path(), ProviderKey, {})

    async def write_provider_keys(self, keys: ProviderKeysStorage) -> ProviderKeysStorage:
        return await write_validated_json_provider_keys(get_provider_keys_index_path(), keys)

    def generate_id(self) -> int:
        return int(time.time() * 1000)

provider_key_storage_util = ProviderKeyStorageUtil()

File: packages/server/src/utils/storage/provider-key-storage.ts
// packages/server/src/utils/storage/provider-key-storage.ts
import { z, ZodError, type ZodTypeAny } from 'zod'
import path from 'node:path'
import fs from 'node:fs/promises'
import { ProviderKeySchema, type ProviderKey } from 'shared/src/schemas/provider-key.schemas'
import { randomUUID } from 'crypto'
import { normalizeToUnixMs } from '../parse-timestamp'

// Define the base directory for storing provider key data
const DATA_DIR = path.resolve(process.cwd(), 'data', 'provider_key_storage')
const PROVIDER_KEYS_FILE_NAME = 'provider_keys.json'

// Schema for the entire storage file: a record of ProviderKeys keyed by their ID
export const ProviderKeysStorageSchema = z.record(z.string(), ProviderKeySchema)
export type ProviderKeysStorage = z.infer<typeof ProviderKeysStorageSchema>

// --- Path Helper ---
function getProviderKeysFilePath(): string {
  return path.join(DATA_DIR, PROVIDER_KEYS_FILE_NAME)
}

// --- Core Read/Write Functions (Common pattern from your other storage utils) ---

async function ensureDirExists(dirPath: string): Promise<void> {
  try {
    await fs.mkdir(dirPath, { recursive: true })
  } catch (error: any) {
    if (error.code !== 'EEXIST') {
      console.error(`Error creating directory ${dirPath}:`, error)
      throw new Error(`Failed to ensure directory exists: ${dirPath}`)
    }
  }
}

async function readValidatedJson<T extends ZodTypeAny>(
  filePath: string,
  schema: T,
  defaultValue: z.infer<T>
): Promise<z.infer<T>> {
  try {
    await ensureDirExists(path.dirname(filePath)) // Ensure parent directory exists
    const fileContent = await fs.readFile(filePath, 'utf-8')
    const jsonData = JSON.parse(fileContent)
    const validationResult = await schema.safeParseAsync(jsonData)

    if (!validationResult.success) {
      console.error(`Zod validation failed reading ${filePath}:`, validationResult.error.errors)
      console.warn(`Returning default value due to validation failure for ${filePath}.`)
      return defaultValue
    }
    return validationResult.data
  } catch (error: any) {
    if (error.code === 'ENOENT') {
      // File doesn't exist, return the default value
      return defaultValue
    }
    console.error(`Error reading or parsing JSON from ${filePath}:`, error)
    throw new Error(`Failed to read/parse JSON file at ${filePath}. Reason: ${error.message}`)
  }
}

async function writeValidatedJson<T extends ZodTypeAny>(
  filePath: string,
  data: unknown, // Accept unknown initially for validation
  schema: T
): Promise<z.infer<T>> {
  try {
    const validationResult = await schema.safeParseAsync(data)
    if (!validationResult.success) {
      console.error(`Zod validation failed before writing to ${filePath}:`, validationResult.error.errors)
      throw new ZodError(validationResult.error.errors)
    }
    const validatedData = validationResult.data

    await ensureDirExists(path.dirname(filePath)) // Ensure directory exists
    const jsonString = JSON.stringify(validatedData, null, 2) // Pretty print
    await fs.writeFile(filePath, jsonString, 'utf-8')

    return validatedData
  } catch (error: any) {
    console.error(`Error writing JSON to ${filePath}:`, error)
    if (error instanceof ZodError) {
      throw error // Re-throw Zod errors
    }
    throw new Error(`Failed to write JSON file at ${filePath}. Reason: ${error.message}`)
  }
}

// --- Specific Data Accessors ---

export const providerKeyStorage = {
  /** Reads the provider keys data file. */
  async readProviderKeys(): Promise<ProviderKeysStorage> {
    return readValidatedJson(getProviderKeysFilePath(), ProviderKeysStorageSchema, {})
  },

  /** Writes the provider keys data file. */
  async writeProviderKeys(keys: ProviderKeysStorage): Promise<ProviderKeysStorage> {
    return writeValidatedJson(getProviderKeysFilePath(), keys, ProviderKeysStorageSchema)
  },

  /** Generates a unique ID for provider keys. */
  generateId: (): number => {
    return normalizeToUnixMs(new Date())
  }
}